test/test_ssl.rb

```diff
@@ -1222,6 +1222,56 @@ def test_get_ephemeral_key
     end
   end
 
+  def test_fallback_scsv
+    ctx_proc = proc { |ctx|
+        ctx.ssl_version = :TLSv1_2
+    }
+    start_server(ctx_proc: ctx_proc) do |port|
+      ctx = OpenSSL::SSL::SSLContext.new fallback_scsv: false
+      ctx.ssl_version = :TLSv1_2
+      # Here is OK
+      # TLS1.2 supported and this is what we ask the first time
+      server_connect(port, ctx)
+    end
+
+    ctx_proc = proc { |ctx|
+      ctx.ssl_version = :TLSv1_1
+    }
+    start_server(ctx_proc: ctx_proc) do |port|
+      ctx = OpenSSL::SSL::SSLContext.new fallback_scsv: true
+      ctx.ssl_version = :TLSv1_1
+      # Here is OK too
+      # TLS1.2 not supported, fallback to TLS1.1 and signaling the fallback
+      # Server doesn't support better, so connection OK
+      server_connect(port, ctx)
+    end
+
+    # Here is not OK
+    # TLS1.2 is supported, fallback to TLS1.1 (downgrade attack) and signaling the fallback
+    # Server support better, so refuse the connection
+    sock1, sock2 = socketpair
+    ctx1 = OpenSSL::SSL::SSLContext.new
+    ctx1.min_version = OpenSSL::SSL::TLS1_1_VERSION
+    ctx1.max_version = OpenSSL::SSL::TLS1_2_VERSION
+    s1 = OpenSSL::SSL::SSLSocket.new sock1, ctx1
+
+    ctx2 = OpenSSL::SSL::SSLContext.new
+    ctx2.enable_fallback_scsv
+    ctx2.ssl_version = :TLSv1_1
+    s2 = OpenSSL::SSL::SSLSocket.new sock2, ctx2
+    t = Thread.new {
+      assert_raise_with_message(OpenSSL::SSL::SSLError, /inappropriate fallback/) {
+        s2.connect
+      }
+    }
+
+    assert_raise_with_message(OpenSSL::SSL::SSLError, /inappropriate fallback/) {
+      s1.accept
+    }
+
+    assert t.join
+  end
```

Comment:
Sockets created by socketpair have to be closed manually to prevent FD leak (which could be an issue when running tests repeatedly for debugging).

Reply:
Fixed